### PR TITLE
Add support for Pushy service

### PR DESCRIPTION
**Pull request by zabolotnov87**
[Pushy](https://pushy.me/) is a highly-reliable push notification gateway, based on [MQTT](https://pushy.me/support#what-is-mqtt) protocol. It allows to deliver notifications reliably to Android devices in China (Google Cloud Messaging and Firebase Cloud Messaging are blocked in China).

### CODE REVIEW

**Comment by hmistry** [comment]
Glad to see this PR adding Pushy into Rpush. We use Pushy as well as Rpush to cover different clients.
Thank you for making this @zabolotnov87. I hope it gets merged.

**Comment by aried3r** [comment]
Hey! Thanks for the PR, from a first look it looks good, but I won't have any time to review it in depth anytime soon. So if someone else wants to take a stab at it, please go ahead!

One thing tho, maybe we should rename the migration to something like `rpush_3_1_0_add_pushy` so we know where it was introduced.

**Comment by zabolotnov87** [comment]
> One thing tho, maybe we should rename the migration to something like rpush_3_1_0_add_pushy so we know where it was introduced

Done!

File: lib/rpush/daemon/pushy/delivery.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+          @notification = notification
+          @batch = batch
+          @pushy_uri = URI.parse("https://api.pushy.me/push?api_key=#{app.api_key}")
+        end
+
+        def perform
+          response = send_request
+          process_response(response)
+        rescue SocketError => error
+          mark_retryable(notification, Time.now + 10.seconds, error)
```
Comment (hmistry) [inline lib/rpush/daemon/pushy/delivery.rb]: Will this create possibility of an endless retry loop every 10 seconds on socket failure? Are there any max count limit somewhere that will terminate the retries and raise error in logs?

**Comment by hmistry** [comment]
Overall I think it looks good... unfortunately I no longer have access to Android devices and Pushy Service to test this PR, sorry.

File: README.md [hunk 2]
```diff
@@ -206,6 +207,31 @@ n.badge = 4
 n.save!
 ```
 
+#### Pushy
+
+[Pushy](https://pushy.me/) is a highly-reliable push notification gateway, based on [MQTT](https://pushy.me/support#what-is-mqtt) protocol. It allows to deliver notifications reliably to Android devices in China (Google Cloud Messaging and Firebase Cloud Messaging are blocked in China).
```
[the remark this answers is not in the document]
Comment (zabolotnov87) [reply README.md]: done

File: README.md [hunk 3]
```diff
@@ -206,6 +207,31 @@ n.badge = 4
...
+
+Note: current implementation does not include [subscriptions](https://pushy.me/docs/android/subscribe-topics) and supports only Android devices.
```
[the remark this answers is not in the document]
Comment (zabolotnov87) [reply README.md]: done

File: lib/generators/templates/rpush_3_1_0_add_pushy.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class Rpush310AddPushy < ActiveRecord::VERSION::MAJOR >= 5 ? ActiveRecord::Migration[5.0] : ActiveRecord::Migration
+  def self.up
+    add_column :rpush_notifications, :external_id, :string, null: true
```
[the remark this answers is not in the document]
Comment (zabolotnov87) [reply lib/generators/templates/rpush_3_1_0_add_pushy.rb]: updated to `external_device_id `

File: lib/rpush/daemon/pushy/delivery.rb (hunk 1 above)
Comment (zabolotnov87) [reply lib/rpush/daemon/pushy/delivery.rb]: There is no any max count limit. All implementations follows this approach, for example, look at https://github.com/rpush/rpush/blob/master/lib/rpush/daemon/gcm/delivery.rb#L23

Comment (hmistry) [reply lib/rpush/daemon/pushy/delivery.rb]: ok

File: lib/generators/templates/rpush_3_1_0_add_pushy.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (hmistry) [reply lib/generators/templates/rpush_3_1_0_add_pushy.rb]: 👍

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmistry) [reply README.md]: 👍

**Comment by hmistry** [comment]
@zabolotnov87 Change commits not pushed yet... presume you'll get to that soon.

File: lib/generators/rpush_migration_generator.rb [hunk 5]
```diff
@@ -44,6 +44,7 @@ def copy_migration
     add_rpush_migration('rpush_2_7_0_updates')
     add_rpush_migration('rpush_3_0_0_updates')
     add_rpush_migration('rpush_3_0_1_updates')
+    add_rpush_migration('add_pushy')
```
Comment (aried3r) [inline lib/generators/rpush_migration_generator.rb]: The migration template has been changed, this should read `rpush_3_1_0_add_pushy`.

File: rpush.gemspec [hunk 6]
```diff
@@ -42,7 +42,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'rubocop'
   s.add_development_dependency 'byebug'
 
-  s.add_development_dependency 'pg'
+  s.add_development_dependency 'pg', '~> 0.21'
```
Comment (aried3r) [inline rpush.gemspec]: I don't think we should limit the `pg` gem here. Maybe within the Appraisals, Rails 5.1.5 and 5-stable seem to have support for pg 1.0. https://github.com/rails/rails/pull/31671#issuecomment-357605227

**Comment by aried3r** [comment]
Hey! I've fixed the builds in master that were failing because of the `pg` gem. I've also incorporated some of the changes you made here, like removing and ignoring the `Gemfile.lock`.

Given that, I don't see why we wouldn't merge and release this soon :)

**Comment by hmistry** [comment]
@aried3r I'd recommend keeping Gemfile.lock in the repo as it makes it easy for new contributors to get a working setup soon and be more encouraged to contribute. Bundler recently made this change for new gems for this very reason. For existing gems it's up to the gem maintainers. See [post](http://bundler.io/blog/2018/01/17/making-gem-development-a-little-better.html).

This shouldn't hold up merging. 😃

**Comment by zabolotnov87** [comment]
green

**Pull request merged**